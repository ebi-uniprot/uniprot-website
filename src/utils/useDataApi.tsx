import React, { useState, useEffect } from 'react';
import fetchData from './fetchData';
import 'regenerator-runtime/runtime';

const useDataApi = (url: string) => {
  const [data, setData] = useState<any>({});

  const getData = async () => {
    const result = await fetchData(url);
    setData(result.data);
  };

  useEffect(() => {
    getData();
  }, [url]);

  return data;
};

export default useDataApi;
