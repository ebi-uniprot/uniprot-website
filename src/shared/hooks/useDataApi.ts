import { useEffect, useReducer, useRef } from 'react';
import axios, {
  AxiosResponse,
  AxiosError,
  AxiosRequestConfig,
  AxiosProgressEvent,
  isCancel,
} from 'axios';
import joinUrl from 'url-join';

import useMessagesDispatch from './useMessagesDispatch';

import fetchData from '../utils/fetchData';
import { addMessage } from '../../messages/state/messagesActions';
import * as logging from '../utils/logging';
import { apiPrefix } from '../config/apiUrls/apiPrefix';

import {
  MessageFormat,
  MessageLevel,
} from '../../messages/types/messagesTypes';
import { Namespace } from '../types/namespaces';
import { UserPreferenceKey } from './useLocalStorage';

export type CustomError = AxiosError<undefined | { messages?: string[] }>;

const invalidFieldMessage = /Invalid fields parameter value '(?<field>[^']*)'/;
const namespacedURL = new RegExp(
  joinUrl(
    apiPrefix,
    `(?<namespace>${Object.values(Namespace).join('|')})`,
    'search'
  )
);

// TODO: possible improvement:
// See https://developers.google.com/web/tools/workbox/modules/workbox-broadcast-update#using_broadcast_update
// Listen for a service worker message saying that there is new data to get
// fresher data from cache and replace the data within this custom hook

// Threshold under which a progress event will not be emitted
// This is in order to avoid unecessary renders
const PROGRESS_THRESHOLD = 2_000; // 2s

export type UseDataAPIState<T> = {
  loading: boolean;
  progress?: number;
  data?: T;
  status?: AxiosResponse['status'];
  statusText?: AxiosResponse['statusText'];
  headers?: AxiosResponse['headers'];
  error?: CustomError;
  redirectedTo?: string;
  url?: string | null;
};

enum ActionType {
  INIT = 'INIT',
  PROGRESS = 'PROGRESS',
  SUCCESS = 'SUCCESS',
  ERROR = 'ERROR',
}

type Action<T> =
  | { type: ActionType.INIT; url?: string }
  | { type: ActionType.PROGRESS; progress: number }
  | {
      type: ActionType.SUCCESS;
      response?: AxiosResponse<T>;
      originalURL?: string;
      progress?: 1;
    }
  | { type: ActionType.ERROR; error: CustomError };

const createReducer =
  <T>() =>
  (state: UseDataAPIState<T>, action: Action<T>): UseDataAPIState<T> => {
    // eslint-disable-next-line default-case
    switch (action.type) {
      case ActionType.INIT:
        return {
          loading: true,
          url: action.url,
        };
      case ActionType.PROGRESS:
        return {
          ...state,
          progress: action.progress,
        };
      case ActionType.SUCCESS:
        // eslint-disable-next-line no-case-declarations
        const newState: UseDataAPIState<T> = {
          ...state,
          loading: false,
          progress: action.progress,
          data: action.response?.data,
          status: action.response?.status,
          statusText: action.response?.statusText,
          headers: action.response?.headers,
        };
        if (
          action.response?.request?.responseURL &&
          action.response.request.responseURL !== action.originalURL
        ) {
          newState.redirectedTo = action.response.request.responseURL;
        } else if (
          // Issue with casing in axios-mock-adapter?
          action.response?.request?.responseUrl &&
          action.response.request.responseUrl !== action.originalURL
        ) {
          newState.redirectedTo = action.response.request.responseUrl;
        }
        return newState;
      case ActionType.ERROR:
        return {
          ...state,
          loading: false,
          progress: undefined,
          status: action.error.response?.status,
          statusText: action.error.response?.statusText,
          headers: action.error.response?.headers,
          error: action.error,
        };
    }
  };

function useDataApi<T>(
  url?: string | null,
  options?: AxiosRequestConfig
): UseDataAPIState<T> {
  const [state, dispatch] = useReducer(createReducer<T>(), { loading: !!url });
  const messagesDispatch = useMessagesDispatch();
  const optionsRef = useRef(options);

  useEffect(() => {
    if (options !== optionsRef.current) {
      optionsRef.current = options;
    }
  }, [options]);

  useEffect(() => {
    // need this variable to ensure state updates don't occur when cancelled/unmounted
    // https://github.com/facebook/react/issues/14369#issuecomment-468267798
    let didCancel = false;

    // we don't require a URL, we just don't need data anymore
    // assume succes with no data
    if (!url) {
      dispatch({ type: ActionType.SUCCESS });
      return;
    }

    dispatch({ type: ActionType.INIT, url });

    // variables to handle cancellation
    // eslint-disable-next-line import/no-named-as-default-member
    const source = axios.CancelToken.source();

    // to keep track of the last time we dispatched a progress action
    let lastProgressDate: number;
    // actual request
    fetchData<T>(url, source.token, {
      ...optionsRef.current,
      onDownloadProgress: (progressEvent: AxiosProgressEvent) => {
        optionsRef.current?.onDownloadProgress?.(progressEvent);
        if (didCancel || progressEvent.progress === undefined) {
          return;
        }
        const now = Date.now();
        // If last time we dispatched a progress action was less than threshold
        if (lastProgressDate && now - lastProgressDate < PROGRESS_THRESHOLD) {
          // skip, don't want to refresh too often
          return;
        }
        dispatch({
          type: ActionType.PROGRESS,
          progress: progressEvent.progress,
        });
        lastProgressDate = now;
      },
    }).then(
      // handle ok
      (response: AxiosResponse<T>) => {
        if (didCancel) {
          return;
        }
        dispatch({
          type: ActionType.SUCCESS,
          response,
          originalURL: url,
          progress: 1,
        });
      },
      // catch error
      (error: CustomError) => {
        // Ignore 404 because it might mean "valid request but no data"
        if (error.response?.status && error.response.status !== 404) {
          logging.error(error, { tags: { origin: 'useDataApi', url } });
        }
        if (isCancel(error) || didCancel) {
          return;
        }
        dispatch({ type: ActionType.ERROR, error });
      }
    );

    // handle unmounting of the hook

    return () => {
      source.cancel();
      didCancel = true;
    };
  }, [url, optionsRef]);

  useEffect(() => {
    // TODO: when refactoring, accept a onError callback to do this from outside
    if (state.status === 400) {
      // Just in case a field is updated in the server and is no longer valid
      try {
        for (const message of state.error?.response?.data?.messages || []) {
          const messageMatch = message.match(invalidFieldMessage);
          const invalidField = messageMatch?.groups?.field;
          if (!messageMatch) {
            continue;
          }
          const nsMatch = state.url?.match(namespacedURL);
          if (!nsMatch?.groups?.namespace) {
            continue;
          }
          const key =
            `table columns for ${nsMatch?.groups?.namespace}` as UserPreferenceKey;
          const faultyArray: string[] = JSON.parse(
            localStorage.getItem(key) as string
          );
          const noFaults = faultyArray.filter(
            (column) =>
              // Clean up any wrong type of value, and the faulty value too
              column && typeof column === 'string' && column !== invalidField
          );
          // If no valid columns then store undefined which will cause the
          // default columns to be used.
          const correctArray = JSON.stringify(
            noFaults.length ? noFaults : undefined
          );
          localStorage.setItem(key, correctArray);
          // Signals to all useLocalStorage hooks in use to rerender
          window.dispatchEvent(
            new StorageEvent('storage', {
              key,
              newValue: correctArray,
              storageArea: window.localStorage,
            })
          );
          return;
        }
      } catch {
        /**/
      }
      const messageContent =
        state.error?.response?.data?.messages?.join(',') || '400 Error';
      messagesDispatch(
        addMessage({
          id: messageContent,
          content: messageContent,
          format: MessageFormat.POP_UP,
          level: MessageLevel.FAILURE,
          displayTime: 5_000,
        })
      );
    }
  }, [messagesDispatch, state.status, state.error, state.url]);

  // when changing the URL, the state is set asynchronously, this is to set it
  // to loading synchronously to avoid using previous data
  if (state.url !== url) {
    return { loading: Boolean(url), url };
  }

  return state;
}

export default useDataApi;
