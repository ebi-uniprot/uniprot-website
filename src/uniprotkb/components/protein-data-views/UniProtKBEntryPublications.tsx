import { FC } from 'react';
import { Loader, Message } from 'franklin-sites';

import useDataApi from '../../../shared/hooks/useDataApi';

import { getPublicationsURL } from '../../../shared/config/apiUrls';
import {
  CitationsAPIModel,
  formatCitationData,
} from '../../../supporting-data/citations/adapters/citationsConverter';

import { MessageLevel } from '../../../messages/types/messagesTypes';
import LiteratureCitation from '../../../supporting-data/citations/components/LiteratureCitation';

const UniProtKBEntryPublications: FC<{
  pubmedIds: string[];
}> = ({ pubmedIds }) => {
  const url = getPublicationsURL(pubmedIds);
  const { loading, data, status, error } = useDataApi<{
    results: CitationsAPIModel[];
  }>(url);

  if (error) {
    return (
      <Message level={MessageLevel.FAILURE}>
        {status}: {error.message}
      </Message>
    );
  }

  if (loading || !data) {
    return <Loader />;
  }

  const { results } = data;
  return (
    <>
      {results &&
        results
          .map((literatureItem) => ({
            ...literatureItem,
            ...formatCitationData(literatureItem.citation),
          }))
          .map(({ citation, statistics, pubmedId, journalInfo }) => (
            <LiteratureCitation
              title={citation.title}
              authors={citation.authors}
              key={`${citation.title}-${citation.citationType}-${citation.journal}`}
              pubmedId={pubmedId}
              statistics={statistics}
              journalInfo={journalInfo}
            />
          ))}
    </>
  );
};

export default UniProtKBEntryPublications;
