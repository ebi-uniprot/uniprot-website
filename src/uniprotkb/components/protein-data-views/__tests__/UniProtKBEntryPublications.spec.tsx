import { BrowserRouter as Router } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import UniProtKBEntryPublications from '../UniProtKBEntryPublications';
import mockUniProtKBEntryPublications from './__mocks__/uniprotKBEntryPublications.json';

import useDataApi from '../../../../shared/hooks/useDataApi';

jest.mock('../../../../shared/hooks/useDataApi', () => jest.fn());

describe('UniProtKBEntryPublications', () => {
  it('Should make a call with pubmed ids and render properly', () => {
    useDataApi.mockImplementation(() => ({
      loading: false,
      data: mockUniProtKBEntryPublications,
    }));
    const pubMedIds = ['123', '456'];
    const { asFragment } = render(
      <Router>
        <UniProtKBEntryPublications pubmedIds={pubMedIds} />
      </Router>
    );
    expect(screen.getByText(/Baumkotter/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should make a call with pubmed ids and fail', () => {
    useDataApi.mockImplementation(() => ({
      error: {
        message: 'There has been an error',
      },
    }));
    const pubMedIds = ['123', '456'];
    const { asFragment } = render(
      <Router>
        <UniProtKBEntryPublications pubmedIds={pubMedIds} />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
