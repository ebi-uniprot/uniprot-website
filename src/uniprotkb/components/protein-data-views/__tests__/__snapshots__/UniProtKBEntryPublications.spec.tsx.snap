// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UniProtKBEntryPublications Should make a call with pubmed ids and fail 1`] = `
<DocumentFragment>
  <div
    class="message message--failure"
    role="status"
  >
    <div
      class="message__side-border"
    />
    <test-file-stub
      height="1.125em"
      width="1.125em"
    />
    <section
      class="message__content"
    >
      <small>
        : There has been an error
      </small>
    </section>
  </div>
</DocumentFragment>
`;

exports[`UniProtKBEntryPublications Should make a call with pubmed ids and render properly 1`] = `
<DocumentFragment>
  <article
    class="inline-publication publication"
  >
    <div
      class="publication__columns"
    >
      <div
        class="publication__columns__main"
      >
        <h5
          class="publication__heading tiny"
        >
          A peptide zipcode sufficient for anterograde transport within amyloid precursor protein.
        </h5>
        <div
          class="publication__authors"
        >
          <a
            href="/uniprotkb?query=lit_author:\\"Satpute-Krishnan P.\\""
          >
            Satpute-Krishnan P.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"DeGiorgis J.A.\\""
          >
            DeGiorgis J.A.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Conley M.P.\\""
          >
            Conley M.P.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Jang M.\\""
          >
            Jang M.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Bearer E.L.\\""
          >
            Bearer E.L.
          </a>
        </div>
        <div
          class="publication__abstract"
        >
          <button
            class="button tertiary"
            type="button"
          >
            View abstract [...]
          </button>
        </div>
      </div>
      <div
        class="publication__columns__side"
      >
        <div
          class="publication__columns__side__item"
        >
          <test-file-stub
            height="2em"
            width="1.875em"
          />
          <ul
            class="no-bullet"
          >
            <li>
              <a
                class="external-link"
                href="https://pubmed.ncbi.nlm.nih.gov/17062754"
                rel="noopener"
                target="_blank"
              >
                PubMed
              </a>
            </li>
            <li>
              <a
                class="external-link"
                href="//europepmc.org/article/MED/17062754"
                rel="noopener"
                target="_blank"
              >
                Europe PMC
              </a>
            </li>
            <li>
              <small>
                <a
                  class="external-link"
                  href="https://dx.doi.org/10.1073/pnas.0607527103"
                  rel="noopener"
                  target="_blank"
                >
                  Proc. Natl. Acad. Sci. U.S.A. 103:16532-16537(
                  <time
                    datetime="2006-01-01T00:00:00.000Z"
                  >
                    2006
                  </time>
                  )
                  <test-file-stub
                    data-testid="external-link-icon"
                    width="12.5"
                  />
                </a>
              </small>
            </li>
          </ul>
        </div>
        <div
          class="publication__columns__side__item"
        >
          <small>
            Mapped to
          </small>
          <div
            class="publication__statistics"
          >
            <a
              href="/uniprotkb?facets=reviewed:true&query=(lit_pubmed:undefined)"
              title="UniProtKB reviewed entries: 2"
            >
              <span
                class="bubble--small colour-pastel-blue"
              >
                2
              </span>
              <test-file-stub
                classname="publication__statistics--reviewed"
                height="1em"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
  </article>
  <article
    class="inline-publication publication"
  >
    <div
      class="publication__columns"
    >
      <div
        class="publication__columns__main"
      >
        <h5
          class="publication__heading tiny"
        >
          Quantitative measurements and modeling of cargo-motor interactions during fast transport in the living axon.
        </h5>
        <div
          class="publication__authors"
        >
          <a
            href="/uniprotkb?query=lit_author:\\"Seamster P.E.\\""
          >
            Seamster P.E.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Loewenberg M.\\""
          >
            Loewenberg M.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Pascal J.\\""
          >
            Pascal J.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Chauviere A.\\""
          >
            Chauviere A.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Gonzales A.\\""
          >
            Gonzales A.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Cristini V.\\""
          >
            Cristini V.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Bearer E.L.\\""
          >
            Bearer E.L.
          </a>
        </div>
        <div
          class="publication__abstract"
        >
          <button
            class="button tertiary"
            type="button"
          >
            View abstract [...]
          </button>
        </div>
      </div>
      <div
        class="publication__columns__side"
      >
        <div
          class="publication__columns__side__item"
        >
          <test-file-stub
            height="2em"
            width="1.875em"
          />
          <ul
            class="no-bullet"
          >
            <li>
              <a
                class="external-link"
                href="https://pubmed.ncbi.nlm.nih.gov/23011729"
                rel="noopener"
                target="_blank"
              >
                PubMed
              </a>
            </li>
            <li>
              <a
                class="external-link"
                href="//europepmc.org/article/MED/23011729"
                rel="noopener"
                target="_blank"
              >
                Europe PMC
              </a>
            </li>
            <li>
              <small>
                <a
                  class="external-link"
                  href="https://dx.doi.org/10.1088/1478-3975/9/5/055005"
                  rel="noopener"
                  target="_blank"
                >
                  Phys. Biol. 9:055005-055005(
                  <time
                    datetime="2012-01-01T00:00:00.000Z"
                  >
                    2012
                  </time>
                  )
                  <test-file-stub
                    data-testid="external-link-icon"
                    width="12.5"
                  />
                </a>
              </small>
            </li>
          </ul>
        </div>
        <div
          class="publication__columns__side__item"
        >
          <small>
            Mapped to
          </small>
          <div
            class="publication__statistics"
          >
            <a
              href="/uniprotkb?facets=reviewed:true&query=(lit_pubmed:undefined)"
              title="UniProtKB reviewed entries: 6"
            >
              <span
                class="bubble--small colour-pastel-blue"
              >
                6
              </span>
              <test-file-stub
                classname="publication__statistics--reviewed"
                height="1em"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
  </article>
  <article
    class="inline-publication publication"
  >
    <div
      class="publication__columns"
    >
      <div
        class="publication__columns__main"
      >
        <h5
          class="publication__heading tiny"
        >
          Amyloid precursor protein dimerization and synaptogenic function depend on copper binding to the growth factor-like domain.
        </h5>
        <div
          class="publication__authors"
        >
          <a
            href="/uniprotkb?query=lit_author:\\"Baumkotter F.\\""
          >
            Baumkotter F.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Schmidt N.\\""
          >
            Schmidt N.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Vargas C.\\""
          >
            Vargas C.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Schilling S.\\""
          >
            Schilling S.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Weber R.\\""
          >
            Weber R.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Wagner K.\\""
          >
            Wagner K.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Fiedler S.\\""
          >
            Fiedler S.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Klug W.\\""
          >
            Klug W.
          </a>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Radzimanowski J.\\""
          >
            Radzimanowski J.
          </a>
          <button
            class="button tertiary"
            type="button"
          >
            [...]
          </button>
          , 
          <a
            href="/uniprotkb?query=lit_author:\\"Kins S.\\""
          >
            Kins S.
          </a>
        </div>
        <div
          class="publication__abstract"
        >
          <button
            class="button tertiary"
            type="button"
          >
            View abstract [...]
          </button>
        </div>
      </div>
      <div
        class="publication__columns__side"
      >
        <div
          class="publication__columns__side__item"
        >
          <test-file-stub
            height="2em"
            width="1.875em"
          />
          <ul
            class="no-bullet"
          >
            <li>
              <a
                class="external-link"
                href="https://pubmed.ncbi.nlm.nih.gov/25122912"
                rel="noopener"
                target="_blank"
              >
                PubMed
              </a>
            </li>
            <li>
              <a
                class="external-link"
                href="//europepmc.org/article/MED/25122912"
                rel="noopener"
                target="_blank"
              >
                Europe PMC
              </a>
            </li>
            <li>
              <small>
                <a
                  class="external-link"
                  href="https://dx.doi.org/10.1523/jneurosci.0180-14.2014"
                  rel="noopener"
                  target="_blank"
                >
                  J. Neurosci. 34:11159-11172(
                  <time
                    datetime="2014-01-01T00:00:00.000Z"
                  >
                    2014
                  </time>
                  )
                  <test-file-stub
                    data-testid="external-link-icon"
                    width="12.5"
                  />
                </a>
              </small>
            </li>
          </ul>
        </div>
        <div
          class="publication__columns__side__item"
        >
          <small>
            Mapped to
          </small>
          <div
            class="publication__statistics"
          >
            <a
              href="/uniprotkb?facets=reviewed:true&query=(lit_pubmed:undefined)"
              title="UniProtKB reviewed entries: 1"
            >
              <span
                class="bubble--small colour-pastel-blue"
              >
                1
              </span>
              <test-file-stub
                classname="publication__statistics--reviewed"
                height="1em"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
